precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,516 @@
+		).Bytes()
+		spenderAddr = context.MsgSender()
+	)
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert amount to sdk type: %v", err)
+	}
+
+	call, err := evmtypes.NewERC20BurnFromCall(
+		bridgeAddrBytes,
+		inputs.Token.Bytes(),
+		spenderAddr.Bytes(),
+		inputs.Amount,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ERC20 burnFrom call: %w", err)
+	}
+
+	_, err = m.evmKeeper.ExecuteContractCall(sdkCtx, call)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute ERC20 burnFrom call: %w", err)
+	}
+
+	assetsUnlocked, err := m.bridgeKeeper.SaveAssetsUnlocked(
+		sdkCtx,
+		inputs.Token.Bytes(),
+		sdkAmount,
+		uint8(inputs.Chain),
+		inputs.Recipient,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send AssetsUnlocked to bridge: %w", err)
+	}
+
+	return assetsUnlocked, nil
+}
+
+func (m *BridgeOutMethod) executeBitcoin(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) (*bridgetypes.AssetsUnlockedEvent, error) {
+	bridgeAddr := sdk.AccAddress(common.HexToAddress(
+		evmtypes.AssetsBridgePrecompileAddress,
+	).Bytes())
+	senderAddr := sdk.AccAddress(context.MsgSender().Bytes())
+
+	authorization, expiration := m.authzKeeper.GetAuthorization(
+		context.SdkCtx(), bridgeAddr.Bytes(), senderAddr.Bytes(), SendMsgURL,
+	)
+	if authorization == nil {
+		return nil, fmt.Errorf("%s authorization type does not exist or is expired for address %s", SendMsgURL, senderAddr)
+	}
+
+	if expiration != nil && expiration.Before(context.SdkCtx().BlockTime()) {
+		return nil, fmt.Errorf("authorization expired at %v", expiration)
+	}
+
+	sendAuth, ok := authorization.(*banktypes.SendAuthorization)
+	if !ok {
+		return nil, fmt.Errorf(
+			"expected authorization to be a %T", banktypes.SendAuthorization{},
+		)
+	}
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert amount: [%w]", err)
+	}
+	coins := sdk.Coins{{Denom: evmtypes.DefaultEVMDenom, Amount: sdkAmount}}
+
+	if err := m.validateAuthorizationLimits(sendAuth, coins); err != nil {
+		return nil, fmt.Errorf("authorization validation failed: %w", err)
+	}
+
+	if err := m.bridgeKeeper.BurnBTC(
+		context.SdkCtx(),
+		senderAddr.Bytes(),
+		sdkAmount,
+	); err != nil {
+		return nil, fmt.Errorf("couldn't burn BTC: %w", err)
+	}
+
+	// now update the authorization to spend for the AssetsBridge
+	msg := banktypes.NewMsgSend(senderAddr.Bytes(), bridgeAddr.Bytes(), coins)
+	resp, err := sendAuth.Accept(context.SdkCtx(), msg)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't update authorization: %w", err)
+	}
+
+	if resp.Updated != nil {
+		err = m.authzKeeper.SaveGrant(context.SdkCtx(), bridgeAddr, senderAddr, resp.Updated, expiration)
+	} else {
+		// Authorization fully consumed, delete it
+		m.authzKeeper.DeleteGrant(context.SdkCtx(), bridgeAddr, senderAddr, SendMsgURL)
```

Comment:
Let's remember that the same authorization instance may hold approvals for the MEZO token. Let's double check the grant is NOT deleted in case BTC allowance is zero but MEZO is not. We should delete the grant only if all tokens/coins are zero.